_posts/2019-04-23-autoregressive-flows/experimenting-with-autoregressive-flows.Rmd

```diff
@@ -0,0 +1,645 @@
+
+`masked_autoregressive_default_template` (now leaving out the `tfb_` as we've entered Python-land) uses `masked_dense` to do what you'd suppose a thus-named function might be doing: construct a dense layer that has part of the weight matrix masked out. How? We'll see after a few Python setup statements.
+
+```{python}
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+tfd = tfp.distributions
+tfb = tfp.bijectors
+tf.enable_eager_execution()
+```
+
+
+The following code snippets are taken from `masked_dense` (in its [current form on master](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/bijectors/masked_autoregressive.py)), and when possible, simplified for better readability, accommodating just the specifics of the chosen example - a toy matrix of shape 2x3:
+
+```{python}
+# construct some toy input data (this line obviously not from the original code)
+inputs = tf.constant(np.arange(1.,7), shape = (2, 3))
+
+# (partly) determine shape of mask from shape of input
+input_depth = tf.compat.dimension_value(inputs.shape.with_rank_at_least(1)[-1])
+num_blocks = input_depth
+num_blocks # 3
+```
+
+Our toy layer should have 4 units:
+
+```{python}
+units = 4
+```
+
+The mask is initialized to all zeros. Considering it will be used to elementwise multiply the weight matrix, we're a bit surprised at its shape (shouldn't it be the other way round?). No worries; all will turn out correct in the end. 
+
+```{python}
+mask = np.zeros([units, input_depth], dtype=tf.float32.as_numpy_dtype())
+mask
+```
+
+```
+array([[0., 0., 0.],
+       [0., 0., 0.],
+       [0., 0., 0.],
+       [0., 0., 0.]], dtype=float32)
+```
+       
+Now to "whitelist" the allowed connections, we have to fill in ones whenever information flow _is_ allowed by the autoregressive property:
+
+```{python}
+def _gen_slices(num_blocks, n_in, n_out):
+  slices = []
+  col = 0
+  d_in = n_in // num_blocks
+  d_in
+  d_out = n_out // num_blocks
+  d_out
+  row = d_out 
+  for _ in range(num_blocks):
+    row_slice = slice(row, None)
+    col_slice = slice(col, col + d_in)
+    slices.append([row_slice, col_slice])
+    col += d_in
+    row += d_out
+  return slices
+
+slices = _gen_slices(num_blocks, input_depth, units)
+for [row_slice, col_slice] in slices:
+  mask[row_slice, col_slice] = 1
+
+mask
+```
+
+```
+array([[0., 0., 0.],
+       [1., 0., 0.],
+       [1., 1., 0.],
+       [1., 1., 1.]], dtype=float32)
+```
+
+Again, does this look mirror-inverted? A transpose fixes shape and logic both:
```

Comment:
I think "A transpose fixes both shape and logic:" would be slightly easier to read.